### PR TITLE
fix(core): allow proper type inference when `ngFor` is used with a `trackBy` function

### LINKED ISSUE #42609: Angular 12 build error - Type '(index: any, item: DropDownItem<any>) => string' is not assignable to type 'TrackByFunction<any>'
# 🐞 Bug report

I upgrade my private library to angular 12 from angular 11.
I start to get error when I build the project.

```
Error: projects/.../default-drop-down-items.component.html:64:50 - error TS2322: Type '(index: any, item: DropDownItem<any>) => string' is not assignable to type 'TrackByFunction<any>'.

64       <ng-container *ngFor="let item of ddItems; trackBy: trackById; let index = index">
                                                    ~~~~~~~

  projects/.../default-drop-down-items.component.ts:31:16
    31   templateUrl: './default-drop-down-items.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component DefaultDropDownItemsComponent.

```

My HTML
```
    <!--Dropdown Items-->
    <ng-container *ngIf="ddItems?.length; else loaderOrEmptySearch">
      <ng-container *ngFor="let item of ddItems; trackBy: trackById; let index = index">
        <ng-container
          *ngTemplateOutlet="itemTemplate;  context: {item: item, index: index }"
        ></ng-container>
      </ng-container>
    </ng-container>
```

My TypeScript Function
```
  trackById(index, item: DropDownItem): string {
    return item.value;
  }
```

Before upgrade to angular 12 the build pass.

In addition when I'm change My TypeScript Function as follow the build work

```
  trackById(index, item: any): string {
    return (item as DropDownItem).value;
  }
```

Am I missing something ?

### Command (mark with an `x`)

- [X] build

### Is this a regression?

Yes, as in angular 11 it worked.

## 🔥 Exception or Error

<pre><code>
Error: projects/.../default-drop-down-items.component.html:64:50 - error TS2322: Type '(index: any, item: DropDownItem<any>) => string' is not assignable to type 'TrackByFunction<any>'.

64       <ng-container *ngFor="let item of ddItems; trackBy: trackById; let index = index">
                                                    ~~~~~~~

  projects/.../default-drop-down-items.component.ts:31:16
    31   templateUrl: './default-drop-down-items.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component DefaultDropDownItemsComponent.

</code></pre>

## 🌍 Your Environment

<pre><code>
Angular CLI: 12.0.5
Node: 14.15.5
Package Manager: npm 6.14.13
OS: darwin x64

Angular: 12.0.5
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, material, platform-browser, platform-browser-dynamic
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1200.5
@angular-devkit/build-angular   12.0.5
@angular-devkit/core            11.2.7
@angular-devkit/schematics      11.2.7
@schematics/angular             12.0.5
ng-packagr                      12.0.6
rxjs                            6.6.7
typescript                      4.2.4

</code></pre>

### PR DESCRIPTION
**Pull request by JoostK**
In #41995 the type of `TrackByFunction` was changed such that the
declaration of a `trackBy` function did not cause the item type to be
widened to the `trackBy`'s item type, which may be a supertype of the
iterated type. This has introduced situations where the template type
checker is now reporting errors for cases where a `trackBy` function is
no longer assignable to `TrackByFunction`.

This commit fixes the error by also including the item type `T` in
addition to the constrained type parameter `U`, allowing TypeScript to
infer an appropriate `T`.

Fixes #42609

### CODE REVIEW

File: packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts [hunk 1]
```diff
@@ -1007,6 +1007,78 @@ export declare class AnimationEvent {
...
+        }
+
+        @NgModule({
+          declarations: [TestCmp],
+          imports: [CommonModule],
+        })
+        class Module {}
+    `);
+
+      env.driveMain();
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:1040]: Would `driveMain()` throw an exception if this didn't compile?
Is it worth doing:

```ts
      const diags = env.driveDiagnostics();
      expect(diags.length).toBe(0);
```

**Review by petebacondarwin — APPROVED** [review]
Nice.

File: packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts (hunk 1 above)
Comment (JoostK) [reply packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:1040]: Yes `driveMain` asserts there are zero diagnostics reported.

**Review by atscott — APPROVED** [review]

**Comment by atscott** [comment]
[presubmit](http://test/OCL:381945279:BASE:381945313:1624915242734:fdc8f8b0)

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: public-api

**Review by atscott — APPROVED** [review]
Reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:384321272:BASE:384321296:1626126300648:5f6872dd).

**Comment by AndrewKushnir** [comment]
@JoostK quick update:
- Regular presubmit was successful
- I've started [a global one](https://test.corp.google.com/OCL:384321272:BASE:384367635:1626142447058:39d08c46) as well to make sure there are no corner-cases that we might've missed.
- It looks like there are some Windows-related CI failures, which might be unrelated to your changes. Could you plz try to rebase the PR to see if that fixes the issue?

**Comment by JoostK** [comment]
The status on the Windows failures is that Paul is looking into it, as we don't have an explanation for why they fail (and how the changes in this PR in particular would affect those tests)
